scanner/string.go

```diff
@@ -0,0 +1,308 @@
+	s.next()
+	s.begin = s.offset
+	c := s.char
+	s.next()
+	switch c {
+	case '@':
+		t, lit := scanAt(s)
+		return s.begin - 1, t, lit
+	case '$':
+		t, lit := scanGlobalVar(s)
+		return s.begin - 1, t, lit
+	case '{':
+		s.pushCtx(stateInsertStmts)
+		return s.begin - 1, token.InsertBegin, nil
+	}
+	return 0, token.Continue, nil
+}
+
+func replace(s *Scanner, c byte, offset int) {
+	s.src[s.offset-offset] = c
+	s.next()
+}
+
+func decodeEscapes(s *Scanner, term byte) (byte, int) {
+	var skip int
+	for s.char != term && s.err == nil {
+		switch s.char {
+		case '#':
+			next := s.peek(2)[1]
+			if isInsertPrefix(next) {
+				return next, skip
+			}
+		case '\\':
+			skip = decodeEscape(s, skip)
+		default:
+			replace(s, s.char, skip)
+		}
+	}
+	return s.char, skip
+}
+
+func decodeEscape(s *Scanner, skip int) int {
+	skip++
+	s.next()
+	if v := escapes[s.char]; v != 0 {
+		replace(s, v, skip)
+		return skip
+	}
+	var n int
+	c := s.char
+	switch c {
+	case '\n':
+		n = 2
+	case '0', '1', '2', '3', '4', '5', '6', '7':
+		n, c = decodeOctalEsc(s)
+	case 'x':
+		skip++
+		s.next()
+		n, c = decodeHexEsc(s)
+	case 'C':
+		skip++
+		s.next()
+		if s.char != '-' {
+			s.failf("invalid escape")
+			return skip
+		}
+		skip++
+		s.next()
+		c = decodeCtrlEsc(s.char)
+	case 'c':
+		skip++
+		s.next()
+		c = decodeCtrlEsc(s.char)
+	}
+	for i := 1; i < n; i++ {
+		skip++
+		s.next()
+	}
+	replace(s, c, skip)
+	return skip
+}
+
+func decodeCtrlEsc(c byte) byte {
+	if c == '?' {
+		return 0x7f
+	}
+	if v := escapes[c]; v != 0 {
+		c = v
+	}
+	return c & 0x9f
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func decodeOctalEsc(s *Scanner) (n int, v byte) {
+	m := min(3, len(s.src)-s.offset)
+	for n = 0; n < m; n++ {
+		c := s.src[s.offset+n]
+		if !token.IsOctadecimal(c) {
+			break
+		}
+		v = v*8 + (c - '0')
+	}
+	if n == 0 {
+		s.failf("invalid octal escape")
+	}
+	return
+}
+
+func decodeHexEsc(s *Scanner) (n int, v byte) {
+	m := min(2, len(s.src)-s.offset)
+	for n = 0; n < m; n++ {
+		c := s.src[s.offset+n]
+		var d byte
+		switch {
+		case '0' <= c && c <= '9':
+			d = c - '0'
+		case 'A' <= c && c <= 'F':
+			d = c - 'A' + 10
+		case 'a' <= c && c <= 'f':
+			d = c - 'a' + 10
+		default:
+			if n == 0 {
+				s.failf("invalid hex escape")
+			}
+			return
+		}
+		v = v*16 + d
+	}
+	return
+}
+
+func stateInsertStmts(s *Scanner) (pos int, t token.Token, literal []byte) {
+	s.begin = s.offset
+	if s.char == '}' {
+		s.next()
+		s.popCtx()
+		return s.begin, token.InsertEnd, nil
+	}
+	return stateCompStmts(s)
+}
+
+func scanSingleQuotedString(s *Scanner, term byte, head int) (token.Token, []byte) {
+	var skip int
+	for s.char != term && s.err == nil {
+		if s.char == '\\' {
+			s.next()
+			if s.char == '\\' || s.char == term {
+				skip++
+			}
+		}
+		replace(s, s.char, skip)
+	}
+	return token.String, s.src[s.begin+head : s.offset-skip]
+}
+
+func scanHeredocBegin(s *Scanner) (token.Token, []byte) {
```

Comment:
__[gometalinter]__
* [gocyclo] cyclomatic complexity 12 of function scanHeredocBegin() is high (> 10)

<!-- Comment by [SideCI](https://sideci.com) -->